server/src/main/scala/org/http4s/server/middleware/authentication/DigestAuth.scala

```diff
@@ -81,11 +81,13 @@ object DigestAuth {
       nonceCleanupInterval: Duration = 1.hour,
       nonceStaleTime: Duration = 1.hour,
       nonceBits: Int = 160,
-  )(implicit F: Async[F]): Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]] = {
-    val nonceKeeper =
-      new NonceKeeper(nonceStaleTime.toMillis, nonceCleanupInterval.toMillis, nonceBits)
-    challenge[F, A](realm, store, nonceKeeper)
-  }
+  )(implicit F: Sync[F]): Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]] =
+    Kleisli { req =>
+      F.delay(new NonceKeeper(nonceStaleTime.toMillis, nonceCleanupInterval.toMillis, nonceBits))
```

Comment:
Hmm, sorry, this doesn't seem quite right either, since it would create a new `NonceKeeper` per-request.

So maybe the signature we want is actually `F[Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]]]`?

Creating the middleware itself is a side-effect.

Reply:
do'h. Yes. Will patch shortly.

Reply:
Wait, that won't work, since `F[Kleisli[F, Request[F], Either[Challenge, AuthedRequest[F, A]]]]` will change the result type of the function

Reply:
OK. I hadn't considered the implications of allocating a new NonceKeeper every request, I'll admit -- I'd gotten a bit too focused on just getting it to compile and the linter to pass. Stepping back, I think the correct approach here is to sacrifice reusability for exposing a new, usable interface to `challenge`, for those that need it.

Consumers of this new function will appreciate the fact that `new NonceKeeper` is suspended in `F.delay` (more predictable tests), while existing users of `DigestAuth.apply` will not be impacted by the change.

Reply:
Oh, woops, I hadn't noticed the `DigestAuth.apply` signature above with the same problem. I wonder if we can deprecate it and fix it too. It's not just about predictable tests, it's really about correctness.

Sorry this is unraveling into more than a small fix 😁 they always do!

Reply:
Not a problem! I'm actually really happy to contribute -- this is just an area I haven't used much, so I don't want to do something haphazardly that breaks usability or ergonomics. I'm usually of the mind that if it's there, it's there for a reason, even though you're right... `F[AuthMiddleware[F, A]]` would be more correct.

Reply:
Given that, how would a backwards-incompatible, type-conflicting change be made? It would be easier if `apply` weren't at fault.

Reply:
Thanks, glad to hear it! Yes, unfortunately I'm not an expert on this auth stuff either.

> I'm usually of the mind that if it's there, it's there for a reason

For sure, but taking a closer look at `NonceKeeper` (and the commit history) I suspect this code has just fallen behind with the times. For example the use of `synchronized` (which might hard-block) where we probably want a semantically-blocking [semaphore](https://typelevel.org/cats-effect/docs/2.x/concurrency/semaphore).
https://github.com/http4s/http4s/blob/61b48b7cfdacd49031c16b8b0d7e293ea43fa8c0/server/src/main/scala/org/http4s/server/middleware/authentication/NonceKeeper.scala#L74-L79

If you're up for it, what we could do here is re-write a modernized, safe `NonceKeeper`, and build the new API around that. But that may be more than you bargained for!

Reply:
Ha! Alright -- I'm alright, if y'all are alright keeping me in line.

To set the stage, I've been in akka-land for the past seven years, even if I was using cats I was never using cats-effect. Don't hold back, but please be patient. I'm here to learn as much as to contribute.

I'll take a crack at it and return in the next few days. Thank you for the guidance thus far!

Reply:
> I've been in akka-land for the past seven years, even if I was using cats I was never using cats-effect.

This was me, one year ago :)